core/src/main/java/org/infinispan/remoting/rpc/RpcManagerImpl.java

```diff
@@ -107,23 +138,164 @@ public String getPendingViewAsString() {
    }
 
    @Override
+   public <T> CompletionStage<T> invokeCommand(Address target, ReplicableCommand command,
+                                               ResponseCollector<T> collector, RpcOptions rpcOptions) {
+      checkTopologyId(command);
+      CacheRpcCommand cacheRpc = toCacheRpcCommand(command);
+
+      if (!statisticsEnabled) {
+         return t.invokeCommand(target, cacheRpc, collector, rpcOptions.deliverOrder(),
+                                rpcOptions.timeout(), rpcOptions.timeUnit());
+      }
+
+      long startTimeNanos = timeService.time();
+      CompletionStage<T> invocation;
+      try {
+         invocation = t.invokeCommand(target, cacheRpc, collector, rpcOptions.deliverOrder(),
+                                      rpcOptions.timeout(), rpcOptions.timeUnit());
+      } catch (Exception e) {
+         return errorReplicating(e);
+      }
+      return invocation.handle((response, throwable) -> updateStatistics(startTimeNanos, response, throwable));
+   }
+
+   private void checkTopologyId(ReplicableCommand command) {
+      if (command instanceof TopologyAffectedCommand && ((TopologyAffectedCommand) command).getTopologyId() < 0) {
+         throw new IllegalArgumentException("Command does not have a topology id");
```

Comment:
It would work, it's just that I don't like `assert` and the fact that it's not active all the time. Too many bad memories running tests without `-ea` from long time ago :)